crates/shared/src/transport/buffered.rs

```diff
@@ -0,0 +1,304 @@
+};
+use futures::{
+    channel::{mpsc, oneshot},
+    future::{self, BoxFuture, FutureExt as _},
+    stream::{self, FusedStream, Stream, StreamExt as _},
+};
+use serde_json::Value;
+use std::{future::Future, sync::Arc, time::Duration};
+use tokio::task::JoinHandle;
+
+/// Buffered transport configuration.
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to send to the node.
+    pub max_concurrent_requests: usize,
+    /// The maximum batch size.
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
+}
+
+impl Default for Configuration {
+    fn default() -> Self {
+        // Default configuration behaves kind of like TCP Nagle.
+        Self {
+            max_concurrent_requests: 1,
+            max_batch_len: MAX_BATCH_SIZE,
+            batch_delay: Duration::default(),
+        }
+    }
+}
+
+/// Buffered `Transport` implementation that implements automatic batching of
+/// JSONRPC requests.
+#[derive(Clone, Debug)]
+pub struct Buffered<Inner> {
+    inner: Arc<Inner>,
+    calls: mpsc::UnboundedSender<CallContext>,
+}
+
+type RpcResult = Result<Value, Web3Error>;
+
+type CallContext = (RequestId, Call, oneshot::Sender<RpcResult>);
+
+impl<Inner> Buffered<Inner>
+where
+    Inner: BatchTransport + Send + Sync + 'static,
+    Inner::Out: Send,
+    Inner::Batch: Send,
+{
+    /// Create a new buffered transport with the default configuration.
+    pub fn new(inner: Inner) -> Self {
+        Self::with_config(inner, Default::default())
+    }
+
+    /// Creates a new buffered transport with the specified configuration.
+    pub fn with_config(inner: Inner, config: Configuration) -> Self {
+        let inner = Arc::new(inner);
+        let (calls, receiver) = mpsc::unbounded();
+        Self::background_worker(inner.clone(), config, receiver);
+
+        Self { inner, calls }
+    }
+
+    /// Start a background worker for handling batched requests.
+    fn background_worker(
+        inner: Arc<Inner>,
+        config: Configuration,
+        calls: mpsc::UnboundedReceiver<CallContext>,
+    ) -> JoinHandle<()> {
+        tokio::task::spawn(batched_for_each(config, calls, move |mut batch| {
+            let inner = inner.clone();
+            async move {
+                match batch.len() {
+                    0 => (),
+                    1 => {
+                        let (id, request, sender) = batch.remove(0);
+                        let result = inner.send(id, request).await;
+                        let _ = sender.send(result);
+                    }
+                    n => {
+                        let (requests, senders): (Vec<_>, Vec<_>) = batch
+                            .into_iter()
+                            .map(|(id, request, sender)| ((id, request), sender))
+                            .unzip();
```

Comment:
Not sure if worth it but we could detect the receiver being dropped through `Sender::is_canceled` and in that case not send a request.

Reply:
Good idea.

> Before if you created a bunch of futures on the transport and did not poll them they wouldn't be executed.

That's a good point. It still behaves slightly differently in the case where you don't drop a request future but just don't poll it (before, it would not do anything, now it still sends the request, even if we check if the `oneshot::Receiver` has been dropped).

Another approach would be to only queue the call once it gets polled. We should still see benefits of batching (we can set a default very small delay - like 1ms or something, to make batching slightly more likely when `join!`-ing futures). WDYT?

Reply:
Depends on how much more complex it gets. Should be easy to either await a small timeout or even await next poll. If it stays easy then it's probably worth it.

Reply:
It was an easy change: https://github.com/gnosis/gp-v2-services/pull/1527/commits/449c3bddca58d9ab12f8237a94b6b12c81d2737a

The behaviour is now:
- Don't send requests for cancelled futures (i.e. dropped `oneshot::Receiver`)
- Only queue the request if it gets polled